### PR TITLE
feat: reporting all base images within Dockerfile

### PR DESCRIPTION
**Pull request by miki725**
- [x] Followed the steps in the contributor's guide: https://crashoverride.com/docs/other/contributing#filing-the-pull-request
- [x] PR title uses [semantic commit messages](https://nitayneeman.com/posts/understanding-semantic-commit-messages-using-git-and-angular/#fix)
- [x] Filled out the template to a useful degree
- [x] Updated `CHANGELOG.md` if necessary

## Issue

currently chalk only reports a single base image

fixes https://github.com/crashappsec/chalk/issues/375

## Description

chalk can report all images used across all of the dockerfile across all of the targets/sections including `COPY` directives

## Testing

```
➜ make tests args="test_docker.py::test_base_images --logs"
```

### CODE REVIEW

File: src/configs/base_keyspecs.c4m [hunk 1]
```diff
@@ -1765,6 +1765,153 @@ Note this only specifies the base image of the final build target.
...
+    },
+    {
+      "from": "busybox:latest",
+      "name": "busybox:latest",
+      "repo": "busybox",
+      "tag": "latest",
+      "src": ["/bin/busybox"],
+      "dest": "bin/busybox",
+    }
+  ],
```
Comment (maximilianhurl) [inline src/configs/base_keyspecs.c4m:1891]: Shouldn't this also have arefence two `ubuntu` as part of two as well?

Comment (miki725) [reply src/configs/base_keyspecs.c4m:1891]: this is for `CHALK_COPY_IMAGES` which reports information about `COPY` directives within each target/section

Comment (maximilianhurl) [reply src/configs/base_keyspecs.c4m:1891]: Ah the text above (https://github.com/crashappsec/chalk/pull/382/files#diff-b93822976ff3c1b7f18f425dc101793754cd64424f682b5599991d37dbf2e0aeR1871) says ``DOCKER_BASE_IMAGES` is going to be:`

Comment (miki725) [reply src/configs/base_keyspecs.c4m:1890]: copy-paste strikes again. fixed some typos. just force pushed fix

thanks for noticing 😄

**Comment by miki725** [comment]
https://app.shortcut.com/crashoverride-1/story/1305/chalk-add-key-for-all-images-referenced-in-dockerfile

**Review by viega — APPROVED** [review]

File: src/docker/dockerfile.nim [hunk 2]
```diff
@@ -982,24 +997,90 @@ proc getTargetDockerSection*(ctx: DockerInvocation): DockerFileSection =
...
-    yield section
+  for s in ctx.getBaseDockerSections(ctx.getTargetDockerSection()):
+    yield s
+
+proc getBaseDockerSection*(ctx: DockerInvocation, section: DockerFileSection):  DockerFileSection =
+  ## get dockerfile section which defines base image
+  ## which is to be pulled from the registry
+  ## (not another section in dockerfile)
+  for s in ctx.getBaseDockerSections(section):
+    return s
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker/dockerfile.nim:1032]: there is no `yield from` in nim as far as I know and the logic is a bit intricate so much safer to reuse some iterators in others vs duplicating DAG loop logic

File: src/configs/base_keyspecs.c4m [hunk 3]
```diff
@@ -1765,6 +1777,159 @@ Note this only specifies the base image of the final build target.
...
+`DOCKER_COPY_IMAGES` is going to be:
+
+```
+{
+  "two": [
+    {
+      "from": "docker",
+      "name": "docker",
+      "repo": "docker",
+      "src": ["/bin/docker"],
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/configs/base_keyspecs.c4m:1898]: test combines both base image and copy statements whereas these examples are per-key. good catch on the typo

**Pull request merged**